src/controller.rs

```diff
@@ -0,0 +1,149 @@
+        draw_screen(&self.stdout, &self.state)
+    }
+
+    pub fn on_startup(&self) -> Result<(), Box<dyn Error>> {
+        self.draw_screen()?;
+        self.tmux_context.prepare()?;
+        Ok(())
+    }
+
+    pub fn on_exit(&self) -> Result<(), Box<dyn Error>> {
+        prepare_screen_for_exit(&self.stdout)
+    }
+
+    pub fn on_keypress_quit(&self) -> Result<(), Box<dyn Error>> {
+        self.tmux_context.cleanup()?;
+        Ok(())
+    }
+
+    pub fn on_keypress_down(&mut self) -> Result<(), Box<dyn Error>> {
+        self.break_pane();
+        self.state.next_process();
+        self.join_pane();
+        self.draw_screen()
+    }
+
+    pub fn on_keypress_up(&mut self) -> Result<(), Box<dyn Error>> {
+        self.break_pane();
+        self.state.previous_process();
+        self.join_pane();
+        self.draw_screen()
+    }
+
+    pub fn on_error(&mut self, err: Box<dyn Error>) {
+        self.state.messages.push(format!("{}", err));
+    }
+
+    pub fn on_keypress_start(&mut self) -> Result<(), Box<dyn Error>> {
+        self.start_process();
+        self.draw_screen()
+    }
+
+    pub fn on_keypress_stop(&mut self) -> Result<(), Box<dyn Error>> {
+        self.halt_process();
+        self.draw_screen()
+    }
+
+    pub fn break_pane(&mut self) {
+        let process = self.state.current_process();
+        if process.pane_status != PaneStatus::Null && process.pane_id.is_some() {
+            self.tmux_context
+                .break_pane(process.pane_id.unwrap(), process.id, &process.label)
+                .unwrap();
+            self.state.set_pane_id(self.state.current_selection, None);
+        }
+    }
+
+    pub fn join_pane(&mut self) {
+        let process = self.state.current_process();
+        if process.pane_status != PaneStatus::Null {
+            let pane_id = self.tmux_context.join_pane(process.id).unwrap();
+            self.state.set_pane_id(self.state.current_selection, Some(pane_id));
+        }
+    }
+
+    pub fn start_process(&mut self) {
+        let process = self.state.current_process().clone();
+
+        if process.status != ProcessStatus::Halted {
+            return;
+        }
+
+        self.state.set_process_running(self.state.current_selection);
+
+        if process.pane_status == PaneStatus::Dead {
+            // TODO: tmux respawn-window
+        }
+
+        if process.pane_status == PaneStatus::Null {
+            let pane_id = self.tmux_context.create_pane(&process.command).unwrap();
+            self.state.set_pane_id(self.state.current_selection, Some(pane_id));
+            self.state.set_pane_running(self.state.current_selection);
+            self.spawn_pid_watcher_thread(
+                self.tmux_context.get_pane_pid(pane_id).unwrap(),
+                self.state.current_selection
+            );
+        }
+    }
+
+    pub fn spawn_pid_watcher_thread(&mut self, pid: i32, process_index: usize) {
+        // spawn(move || unsafe {
```

Comment:
I'm not sure how to get this block of code working yet. I had it working before I added the controller using Arc, and I thought adding the controller would help make it more apparent how to write this. I haven't spent much time on it again yet, but if you have ideas/knowledge to share that'd be great.